EpikCord/client/http.py

```diff
@@ -0,0 +1,296 @@
+        files: Optional[List[File]] = None,
+        **kwargs,
+    ) -> aiohttp.ClientResponse:
+        """
+        Parameters
+        ----------
+        *args
+            The args to pass to the ClientSession.request method.
+        discord: bool
+            Whether this request is to Discord.
+        json: Optional[Dict[:class:`str`, Any]]
+            The json document to pass to the ClientSession.request method.
+        files: Optional[List[File]]
+            The files to pass to the request.
+        **kwargs
+            The kwargs to pass to the ClientSession.request method.
+
+        Returns
+        -------
+        aiohttp.ClientResponse
+            The response of the request.
+
+        Raises
+        ------
+        NotFound
+            If the request returns a 404 status code.
+        Forbidden
+            If the request returns a 403 status code.
+        Unauthorized
+            If the request returns a 401 status code.
+        BadRequest
+            If the request returns a 400 status code.
+        HTTPException
+            If the request returns a status code that is not OK and not any of the other exceptions.
+        """
+
+        method = route.method
+        url = route.url
+
+        if not discord:
+            return await self.session.request(method, url, *args, **kwargs)
+
+        self.setup_kwargs(kwargs, files=files, json=json)
+
+        url = self.clean_url(url)
+        bucket = self.buckets.get(f"{method}:{url}") or MockBucket()
+
+        await self.global_event.wait()
+        await bucket.wait()
+
+        for _ in range(5):
+            async with self.session.request(method, url, *args, **kwargs) as response:
+                if isinstance(bucket, MockBucket):
+                    bucket = await self.set_bucket(
+                        response=response,
+                        **route.major_parameters,
+                    )
+
+                data = await self.extract_content(response)
+
+                await self.log_request(response, data)
+
+                if response.status == 429:
+                    await self.handle_ratelimit(data, bucket)
+                elif response.headers.get("X-RateLimit-Remaining", "1") == "0":
+                    await self.handle_exhausted_bucket(response, bucket)
+                elif not response.ok:
+                    error = self.error_mapping.get(response.status, HTTPException)
+                    raise error(response, data)
+                elif response.ok:
+                    return response
+        raise TooManyRetries(f"Attempted {method} request to {url} 5 times but failed.")
+
+    async def set_bucket(
+        self,
+        *,
+        response: aiohttp.ClientResponse,
+        channel_id: Optional[int] = None,
+        guild_id: Optional[int] = None,
+        webhook_id: Optional[int] = None,
+        webhook_token: Optional[str] = None,
+    ) -> Union[Bucket, TopLevelBucket]:
```

Comment:
Why all of this is needed by the function? Couldn't this be improved?

Reply:
They're all major parameters that are needed for rate limit handling.

Reply:
> They're all major parameters that are needed for rate limit handling.

Then make a object for this.